File: models/pos_order.py
# -*- coding: utf-8 -*-
from odoo import models, fields, api
from odoo.exceptions import ValidationError


class PosOrder(models.Model):
    _inherit = 'pos.order'
    
    @api.constrains('partner_id', 'config_id')
    def _check_partner_required(self):
        """Valida que exista un cliente si la configuración lo requiere"""
        for order in self:
            if order.config_id.require_customer and not order.partner_id:
                raise ValidationError(
                    'Debe seleccionar un cliente para completar esta orden. '
                    'La configuración del PDV "%s" requiere un cliente obligatorio.' % order.config_id.name
                )
    
    @api.model
    def _order_fields(self, ui_order):
        """Asigna el cliente predeterminado si no hay uno seleccionado"""
        order_fields = super(PosOrder, self)._order_fields(ui_order)
        
        # Si no hay cliente en la orden y el PDV tiene un cliente predeterminado
        if not order_fields.get('partner_id'):
            pos_config = self.env['pos.config'].browse(ui_order.get('pos_session_id'))
            if pos_config and pos_config.default_customer_id:
                order_fields['partner_id'] = pos_config.default_customer_id.id
        
        return order_fields

File: static/src/app/models/pos_order.js
/** @odoo-module */

import { PosOrder } from "@point_of_sale/app/models/pos_order";
import { patch } from "@web/core/utils/patch";

patch(PosOrder.prototype, {
    setup(vals) {
        super.setup(...arguments);
        
        // Asignar cliente predeterminado al crear la orden
        setTimeout(() => {
            this.assignDefaultCustomerIfNeeded();
        }, 100);
    },
    
    assignDefaultCustomerIfNeeded() {
        // Solo si no hay cliente ya
        if (this.partner_id) {
            return;
        }
        
        // Verificar configuración
        const config = this.session?.config_id;
        if (!config?.default_customer_id) {
            return;
        }
        
        // default_customer_id ya ES el partner, no un ID
        this.update({ partner_id: config.default_customer_id });
    },
});